log: Overhaul Windows Logger with Pipes and FHs

johnwparent opened the pull request:
Redo how we handle IO direction for logging on Windows

Previous we closed sys.stdout, replaced it with a new buffer pointing at the logging file, and then dup2'd to redirect stdout. The logfile was then read from a thread, controlled by the main spack thread, and said reader thread forwarded on the logging context to main Spack's dup'd stdout.

This is deeply insufficient for a number of use cases:

It did not support filtering, unicode, control characters, or force_echo

All echo, including force_echo when supported, was racey, due to the disparate mechanisms by which reader thread behavior was controlled vs the context from which it obtained information. I.e. the reader thread could be killed or echo turned off, before any stdout buffers could be flushed (or whatever out-of-sync ordering) and the reader thread wouldn't see the log output it was supposed to echo prior to turning echo off or terminating.

Nested logging, or cached references to stdout and stderr were invalidated. Previously, on io redirection, we were closing stdout and replacing it with our new buffer. On io restoration, we were restored a buffer pointing at console stdout to sys.stdout, but any cached references to stdout (i.e. nested logging contexts or pytest) were invalidated.

Now we:

* do not close whatever sys.stdout refers to, we mearly store a reference to it
* do not construct a new buffer from python's api, instead we perform redirection from the ground up, starting with fhs to fds and compose a buffer from the existing fh's fd.
* Use pipes for communication with our reader thread, which enables filtering, control characters, etc and allows for non racey echo behavior.

At a glance:

* Redirects IO starting at the Windows file handle (fh) level
* Properly maps fhs to fds, and obtains the appropriate buffer wrapping each
* Enables proper subprocess io redirection inheritance
* Manages shared file handles between stderr and stdout on redirect
* Properly redirects IO at each level of handling, kernel, C, and python
* Ensures cached references to console streams are not invalidated
* Ensures correct order of release for buffers -> fhs

Moves toward a unified logging class by introducing fundamental support
 for the concepts leveraged by the non Windows logger classes like pipes


Also enabled logging testing for Windows

Adds a test for nested logging context managers.

johnwparent committed:
log: Overhaul Windows Logger with Pipes and FHs

Redo how we handle IO direction for logging on Windows

Previous we closed sys.stdout, replaced it with a new buffer pointing
at the logging file, and then dup2'd to redirect stdout.
The logfile was then read from a thread, controlled by the main spack
thread, and said thread forwarded on the logging context to main Spack's
dup'd stdout.

This is deeply insufficient for a number of use cases:

It did not support filtering, unicode, control characters, or force_echo

All echo, including force_echo if it did support, was racey, due to the
disparate mechanism for which the reader thread had to control thread behavior
vs where it was obtaining information. I.e. the reader thread could be
killed or echo turned off from pyhon, before any stdout buffers could be flushed
and the reader thread wouldn't see the log output it was supposed to echo before
turning echo off or terminating.

Nested logging, or cached references to stdout and stderr were invalidated
Previously, on io redirection, we were closing stdout and replacing it with
our new buffer. On io restoration, we were restored a buffer pointing at console stdout
to sys.stdout, but any cached references to stdout (i.e. nested logging contexts or pytest)
were invalidated.

Now we:

* do not close whatever sys.stdout refers to, we mearly store a reference to it
* do not construct a new buffer from python's api, instead we perform redirection from
the ground up, starting with fhs to fds and compose a buffer from the existing fh's fd.
* Use pipes for communication with our reader thread, which enables filtering, control characters, etc
and allows for non racey echo behavior.

At a glance:

* Redirects IO starting at the Windows file handle (fh) level
* Properly maps fhs to fds, and obtains the appropriate buffer wrapping each
* Enables proper subprocess io redirection inheritance
* Manages shared file handles between stderr and stdout on redirect
* Properly redirects IO at each level of handling, kernel, C, and python
* Ensures cached references to console streams are not invalidated
* Ensures correct order of release for buffers -> fhs

Moves toward a unified logging class by introducing fundamental support
 for the concepts leveraged by the non Windows logger classes like pipes

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Can't annotate type only avail on Windows

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
No dup_fh on non Windows

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
filter_fn is optional

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
windows only method

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
How did this pass style?

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Import fixes

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
sigh, mypy

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
io isolation in new logging test

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
+                # Flush the system stream before redirection
+                sys_stream.flush()
+        except BaseException:
+            # swallow flush errors
```
This would indicate mismanagement wouldn't it? We might want to log this at the debug level.

johnwparent replied:
Yes, this would indicate an incorrectly timed flush, added debug logging.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
-        # Flush and close sys_stream - also closes the file descriptor (fd)
-        sys_stream = getattr(sys, self.sys_attr)
-        sys_stream.flush()
-        sys_stream.close()
```
Does this address the case where a context-manager-like functionality above this has saved a reference to sys.stdout?

johnwparent replied:
Yes, we're only closing IO components owned by this logging redirect instance.

haampie commented:
I haven't looked at this in detail, but just wondering if you've seen `class Tee` in `new_installer.py` as well? The `log_output` bits will likely go away with the old installer.

johnwparent commented:
> I haven't looked at this in detail, but just wondering if you've seen `class Tee` in `new_installer.py` as well? The `log_output` bits will likely go away with the old installer.

Ahh, I have not. Perhaps this PR should be refactored in favor of supporting `Tee` on Windows then. I'm away at the moment, but I'll look at it once I'm back.

Thanks for pointing that out.

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -165,4 +160,21 @@ def test_log_subproc_and_echo_output(capfd, tmp_path: pathlib.Path):
         # Check captured output (echoed content)
         # Note: 'logged' is not echoed because force_echo() scope ended
-        assert capfd.readouterr()[0] == "echo\n"
+        # "print(echo)" above automatically uses an "\r\n" on Windows
```
(minor) add Note: or something to indicate this is not a continuation of prior note

johnwparent replied:
Oh yeah good point. Thanks.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
+        dup_redirect_h = dup_fh(redirect_h)
+        # ensure inheritable handles
+        os.set_handle_inheritable(redirect_h, True)
+        # Todo: investigate if this is neccesary if we do the above before the handle dup
```
Question(s): Can this be determined before a merge? I think redirect_h is what's being used by subprocs, and I would have expected that's the only one that needs to be inheritable.

johnwparent replied:
Investigating now.

johnwparent replied:
They both need to be inheritable as they are both involved in redirection. Because one is managed by the OS, and one by python/c, we need both handles to do the redirection and avoid a double close on a singular handle.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
+            raise KeyError(self.sys_attr)
+
+        self.saved_stream = getattr(sys, self.sys_attr)
+        self.saved_std_fd = self.saved_stream.fileno()
```
Do we need to track this explicitly? It seems to always be available via `self.saved_stream.fileno()`

If I'm right but you still think it should be tracked, I can think on that. I worry about situations where duplicate data is maintained with separate references though.

johnwparent replied:
I think it's safer to track it explicitly yes. Basically it allows us to have a consistent idea of what fd we're using to replace whatever the std stream happens to be at redirection time without needing to consider the state of saved_stream.

scheibelp replied:
Maybe renaming this might help: calling it `saved_std_fd` implies to me there is some different FD, but really we are `dup`ing onto that FD and the number stays the same the whole time, so maybe just `std_fd` (plus some sort of comment to that effect) would be helpful.

johnwparent replied:
Makes sense to me.

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -165,4 +160,21 @@ def test_log_subproc_and_echo_output(capfd, tmp_path: pathlib.Path):
+        # "print(echo)" above automatically uses an "\r\n" on Windows
+        # and will replace any \n with \r\n (so end=\n does not work)
+        # \r\n is expected and correct here
+        newline = "\r\n" if sys.platform == "win32" else "\n"
```
It looks like redirect_stream is setting `newline="\n",`, so does that not have any effect?

johnwparent replied:
It's been a little while but IIRC this is an artifact of pytest and capfd. If you look at the check above where we are able to use `\n` consistently in the file output, that is the effect of setting newline in redirect_stream.

scheibelp replied:
It seems to me that depending on the context, the newline format *might* not be `\r\n` on Windows, do you think it would make sense to allow for either line ending on Windows? I think the point is that whatever the line ending ends up being is up to whatever configured stdout.

johnwparent replied:
> It seems to me that depending on the context, the newline format might not be \r\n on Windows

Very correct. 

>  think the point is that whatever the line ending ends up being is up to whatever configured stdout.

There are actually a few cases where opening files with unknown line endings with `\r\n` can cause double newlines (`\r\n\n`) so its often safer in general for us to open streams with `\n`. 

In general, `newlines` is impacting the translation at write and read time through the python API. So to a degree, it doesn't matter what we do, the caller can specify the line ending they want when they open whatever stream they're interested in for reading. 

Lets say someone has stdout opened for `\r\n`, and we redirect it with our logger. If they were to capture our redirected (and tee'd) stdout, they'd need another `open` call, which they'd be able to specify their encoding with. 
The only case in which `\n` wouldn't be translated to `\r\n` is if they specified an explicit translation, or explicitly turned translation off, in which case the user has made a deliberate choice. 

We can see that here, where in most of the cases of these log tests, we're able to consistently and nicely use `\n`, except with capfd, where `\r\n` has been enforced by the capfd mechanism and our logger does not overwrite that.

We also definitely want any files we're working with to be `\n`. 

Opening our pipe as `\n` is I think the right choice as we're the only readers and we don't bother dealing with newlines directly, but using one line ending everywhere that every single modern program understands allows spack to have more generic, portable cross platform code, and is one fewer place we have to deal with line endings. I'm all for forcing Windows to use `\n`.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -630,68 +654,107 @@ def __init__(
+            read.close()
+            log_writer.close()
+            stdout.flush()
+            stdout.close()
```
I don't think you should have to flush before you close.

johnwparent replied:
You are correct, removed.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -630,68 +654,107 @@ def __init__(
+                    # the pipe is closed or otherwise inaccesible
+                    return
+                norm_data = data.decode(encoding="utf-8", errors="replace")
+                # stdout.write(data.decode())
```
vestigial?

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -630,68 +654,107 @@ def __init__(
+                    output = clean_line
+                    if filter_fn:
+                        output = filter_fn(output)
+                    enc = stdout.encoding
```
(minor) Can this check be done outside the loop (and e.g. define a lambda here?

Might be worth printing a warning if the encoding is not utf-8

Also question: is there a reason you can't consume the bytes from the pipe and pass them directly through without detecting/matching encoding?

johnwparent replied:
Windows love utf-16 (wide), so depending what stdout is, it may be utf-16, which is normal and expected, particularly if we're in nested logging contexts and stdout is a pipe. 

> Also question: is there a reason you can't consume the bytes from the pipe and pass them directly through without detecting/matching encoding?

The encoding we're reading may not match the encoding of the output, without this output is often garbled/unreadable (or mandarin characters, I hit those a lot without this)

scheibelp replied:
For purposes of echoing, I agree there has to be decoding of the reader bytes, and encoding when passing to stdout (although possibly not when it's not a tty)

For writing to the log file, could the bytes just be passed through?

i.e. for the `log_writer`, can we just

```
log_writer.write(data)
```

vs.

```
norm_data = data.decode(encoding="utf-8", errors="replace")
clean_line, num_controls = control.subn("", norm_data)
log_writer.write(_strip(clean_line).encode(encoding="utf-8"))
```

johnwparent replied:
We still need to process the data for control characters, so we still need to decode (in utf-8), to get `clean_line` (devoid of control characters, cruft, etc), and clean line needs to be encoded back to binary to be written to the out pipe for the log (and stdout). So all we're doing "extra" is ensuring a consistent utf-8 format for the log.

Looking a bit lower, we actually do encode the output intended for stdout as whatever the terminal happens to be.

That makes sense for stdout, but for the logs, which are spack controlled entities, we should probably consistently make those utf-8 right?

scheibelp reviewed (commented):
Not complete, but I have a few preliminary requests and questions

scheibelp reviewed (commented):
Not complete, but I have some preliminary requests/questions

johnwparent committed:
wip

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Quick review updates

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Loosen Checks for MacOS in CI

MacOS GH CI tests have some issue with coverage not finding the "bin" module that also does not occur with local testing on mac, linux, or Windows. Loosen the check so the extra error cruft of converage does not cause the test to fail, and we still check what we need to check

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -547,67 +554,79 @@ class StreamWrapper:
+        redirect_h = write_conn.fileno()
+        dup_redirect_h = dup_fh(redirect_h)
+        os.set_handle_inheritable(redirect_h, True)
+        os.set_handle_inheritable(dup_redirect_h, True)
```
Continuing from https://github.com/spack/spack/pull/51791#discussion_r3061209036 (which I have collapsed because it's out of date)

If I comment out this, then `spack unit-test -k log.py` still passes. If I comment out both `set_handle_inheritable` lines, then one test fails. Should the tests be expanded to cover this. I can say at the moment I'm having trouble understanding the need and that a test might help.

johnwparent replied:
I think this was just my mistake, you're right, one does not need to be inherited. 

For full context, I observed similar behavior, but I was thinking there may be an untested edge case where we'd want both inheritable (in which case your request for more testing is right on the money), but I don't think the edge case I'm thinking of is reasonable/possible.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -630,68 +649,104 @@ def __init__(
+    ):
+        force_echo = False
+
+        # Open both write and reading on logfile
```
Why is reading mode also enabled? I only ever see it written (andI don't think we should be reading from it).

johnwparent replied:
I think this comment is vestigial, we used to be doing both, but now we only write since we're closer to doing actual tee behavior.

johnwparent committed:
Small updates

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Log write only

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Document line ending test behavior

Signed-off-by: John Parent <john.parent@kitware.com>
